kube-runtime/src/utils/stream_subscribe.rs

```diff
@@ -0,0 +1,236 @@
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Arc::new(Ok(3)))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Arc::new(Ok(4)))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(None));
+    }
+
+    #[tokio::test]
+    async fn all_subscribers_get_events() {
+        let events = [Ok(0), Ok(1), Err(2), Ok(3), Ok(4)];
+        let rx = stream::iter(events);
+        let rx = StreamSubscribe::new(rx);
+
+        let rx_s1 = rx.subscribe();
+        let rx_s2 = rx.subscribe();
+
+        pin_mut!(rx);
+        pin_mut!(rx_s1);
+        pin_mut!(rx_s2);
+
+        // Subscribers are pending until we start consuming the stream
+        assert_eq!(poll!(rx_s1.next()), Poll::Pending, "rx_s1");
+        assert_eq!(poll!(rx_s2.next()), Poll::Pending, "rx_s2");
+
+        for item in events {
+            assert_eq!(poll!(rx.next()), Poll::Ready(Some(Arc::new(item))), "rx");
+            let expected = Poll::Ready(Some(Ok(Arc::new(item))));
+            assert_eq!(poll!(rx_s1.next()), expected, "rx_s1");
+            assert_eq!(poll!(rx_s2.next()), expected, "rx_s2");
+        }
+
+        // Ensure that if the stream is closed, all subscribers are closed
+        assert_eq!(poll!(rx.next()), Poll::Ready(None), "rx");
+        assert_eq!(poll!(rx_s1.next()), Poll::Ready(None), "rx_s1");
+        assert_eq!(poll!(rx_s2.next()), Poll::Ready(None), "rx_s2");
+    }
+
+    #[tokio::test]
+    async fn subscribers_can_catch_up_to_the_main_stream() {
+        let events = (0..CHANNEL_CAPACITY).map(Ok::<_, ()>).collect::<Vec<_>>();
+        let rx = stream::iter(events.clone());
+        let rx = StreamSubscribe::new(rx);
+
+        let rx_s1 = rx.subscribe();
+
+        pin_mut!(rx);
+        pin_mut!(rx_s1);
+
+        for item in events.clone() {
+            assert_eq!(poll!(rx.next()), Poll::Ready(Some(Arc::new(item))), "rx",);
+        }
+
+        for item in events {
+            assert_eq!(
+                poll!(rx_s1.next()),
+                Poll::Ready(Some(Ok(Arc::new(item)))),
+                "rx_s1"
+            );
+        }
+    }
+
+    #[tokio::test]
+    async fn if_the_subscribers_lag_they_get_a_lagged_error_as_the_next_event() {
+        // The broadcast channel rounds the capacity up to the next power of two.
+        let max_capacity = CHANNEL_CAPACITY.next_power_of_two();
+        let overflow = 5;
+        let events = (0..max_capacity + overflow).collect::<Vec<_>>();
+        let rx = stream::iter(events.clone());
+        let rx = StreamSubscribe::new(rx);
+
+        let rx_s1 = rx.subscribe();
+
+        pin_mut!(rx);
+        pin_mut!(rx_s1);
+
+        // Consume the entire stream, overflowing the inner channel
+        for _ in events {
+            rx.next().await;
+        }
+
+        assert_eq!(
+            poll!(rx_s1.next()),
+            Poll::Ready(Some(Err(Error::Lagged(overflow as u64)))),
+        );
+
+        let expected_next_event = overflow;
```

Comment:
so we've gone al the way to the end to value 128+5 thus erasing the first 5.
good test, and nicely documented in the interface 👍